managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerTest.java

```diff
@@ -291,6 +291,185 @@ public void acknowledge1() throws Exception {
+                                ManagedLedger ledger = pair.getLeft();
+                                ManagedCursor cursor = pair.getRight();
+                                assertEquals(((ManagedLedgerImpl) ledger).getCacheSize(), message1.getBytes(Encoding).length);
+
+                                cursor.asyncReadEntries(1, new ReadEntriesCallback() {
+                                    @Override
+                                    public void readEntriesComplete(List<Entry> entries, Object ctx) {
+                                        ManagedCursor cursor = (ManagedCursor) ctx;
+                                        assertEquals(entries.size(), 1);
+                                        Entry entry = entries.get(0);
+                                        final Position position = entry.getPosition();
+                                        assertEquals(new String(entry.getDataAndRelease(), Encoding), message1);
+                                        ((ManagedLedgerImpl) ledger).doCacheEviction(
+                                                System.nanoTime() - TimeUnit.MILLISECONDS.toNanos(30000));
+                                        assertEquals(((ManagedLedgerImpl) ledger).getCacheSize(), message1.getBytes(Encoding).length);
+
+                                        log.debug("Mark-Deleting to position {}", position);
+                                        cursor.asyncMarkDelete(position, new MarkDeleteCallback() {
+                                            @Override
+                                            public void markDeleteComplete(Object ctx) {
+                                                log.debug("Mark delete complete");
+                                                ManagedCursor cursor = (ManagedCursor) ctx;
+                                                assertFalse(cursor.hasMoreEntries());
+                                                ((ManagedLedgerImpl) ledger).doCacheEviction(
+                                                        System.nanoTime() - TimeUnit.MILLISECONDS.toNanos(30000));
+                                                assertEquals(((ManagedLedgerImpl) ledger).getCacheSize(), 0);
+
+                                                counter.countDown();
+                                            }
+
+                                            @Override
+                                            public void markDeleteFailed(ManagedLedgerException exception, Object ctx) {
+                                                fail(exception.getMessage());
+                                            }
+
+                                        }, cursor);
+                                    }
+
+                                    @Override
+                                    public void readEntriesFailed(ManagedLedgerException exception, Object ctx) {
+                                        fail(exception.getMessage());
+                                    }
```

Comment:
I am not sure that calling 'fail' not in the main TestNG method/thread will cause the test to fail

Reply:
I call 'fail' directly locally, which can cause the test to fail.
I have tried the following ways:
public void testCacheEvictionByMarkDeletedPosition() throws Throwable{
fail("xxx");
.....
}

public void testCacheEvictionByMarkDeletedPosition() throws Throwable{
.....
public void addComplete(Position position, ByteBuf entryData, Object ctx) {
fail("xxx");
......
}
}

Reply:
Fixed: use CompletableFuture 
PTAL,thanks! @eolivelli